src/Storages/Cache/ExternalDataSourceCache.cpp

```diff
@@ -0,0 +1,258 @@
+}
+
+
+RemoteReadBuffer::RemoteReadBuffer(size_t buff_size) : BufferWithOwnMemory<SeekableReadBufferWithSize>(buff_size)
+{
+}
+
+RemoteReadBuffer::~RemoteReadBuffer()
+{
+    if (file_cache_controller)
+        file_cache_controller->deallocFile(std::move(file_buffer));
+}
+
+std::unique_ptr<ReadBuffer> RemoteReadBuffer::create(ContextPtr context, IRemoteFileMetadataPtr remote_file_metadata, std::unique_ptr<ReadBuffer> read_buffer)
+{
+    auto * log = &Poco::Logger::get("RemoteReadBuffer");
+    size_t buff_size = DBMS_DEFAULT_BUFFER_SIZE;
+    if (read_buffer)
+        buff_size = read_buffer->internalBuffer().size();
+    /*
+     * in the new implement of ReadBufferFromHDFS, buffer size is 0.
+     *
+     * in the common case, we don't read bytes from readbuffer directly, so set buff_size = DBMS_DEFAULT_BUFFER_SIZE
+     * is OK.
+     *
+     * we need be careful with the case  without local file reader.
+     */
+    if (buff_size == 0)
+        buff_size = DBMS_DEFAULT_BUFFER_SIZE;
+
+    auto remote_path = remote_file_metadata->remote_path;
+    auto remote_read_buffer = std::make_unique<RemoteReadBuffer>(buff_size);
+    ErrorCodes::ErrorCode error;
+
+    std::tie(remote_read_buffer->file_cache_controller, read_buffer, error) = ExternalDataSourceCache::instance().createReader(context, remote_file_metadata, read_buffer);
+    if (remote_read_buffer->file_cache_controller == nullptr)
+    {
+        LOG_ERROR(log, "Failed to allocate local file for remote path: {}, reason: {}.", remote_path, error);
+        // read_buffer is the input one.
+        return read_buffer;
+    }
+    else
+    {
+        remote_read_buffer->file_buffer = remote_read_buffer->file_cache_controller->allocFile();
+        if (!remote_read_buffer->file_buffer)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Create file readbuffer failed. {}",
+                    remote_read_buffer->file_cache_controller->getLocalPath().string());
+    }
+    remote_read_buffer->remote_file_size = remote_file_metadata->file_size;
+    return remote_read_buffer;
+}
+
+bool RemoteReadBuffer::nextImpl()
+{
+    auto start_offset = file_buffer->getPosition();
+    auto end_offset = start_offset + file_buffer->internalBuffer().size();
+    file_cache_controller->waitMoreData(start_offset, end_offset);
+
+    auto status = file_buffer->next();
+    if (status)
+    {
+        BufferBase::set(file_buffer->buffer().begin(),
+                file_buffer->buffer().size(),
+                file_buffer->offset());
+        ProfileEvents::increment(ProfileEvents::ExternalDataSourceLocalCacheReadBytes, file_buffer->available());
+    }
+    return status;
+}
+
+off_t RemoteReadBuffer::seek(off_t offset, int whence)
+{
+    if (!file_buffer)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot call seek() in this buffer. It's a bug!");
+    /*
+     * Need to wait here. For example, the current file has been download at position X, but here we try to seek to
```

Comment:
btw for example with parquet format it will seek to the end of file to read file metadata and then seek to start to do actual reading. But the file is downloaded into cache from start to end sequentially. So we will have to wait until it is fully downloaded. Do you think it is ok? I have some doubts.